LibreHardwareMonitorLib/Hardware/Battery/BatteryGroup.cs

```diff
@@ -0,0 +1,231 @@
+        private static bool IsValue(IntPtr ptr) => ptr.ToInt64() >> 16 == 0;
+
+        private static string GetString(IntPtr ptr, CharSet charSet = CharSet.Auto, long allocatedBytes = long.MaxValue)
+        {
+            if (IsValue(ptr)) return null;
+            var sb = new StringBuilder();
+            unsafe
+            {
+                var chkLen = 0L;
+                if (GetCharSize(charSet) == 1)
+                {
+                    for (var uptr = (byte*)ptr; chkLen < allocatedBytes && *uptr != 0; chkLen++, uptr++)
+                        sb.Append((char)*uptr);
+                }
+                else
+                {
+                    for (var uptr = (ushort*)ptr; chkLen + 2 <= allocatedBytes && *uptr != 0; chkLen += 2, uptr++)
+                        sb.Append((char)*uptr);
+                }
+            }
+            return sb.ToString();
+        }
+
+        public BatteryGroup(ISettings settings)
+        {
+            // No implementation for battery information on Unix systems
+            if (Software.OperatingSystem.IsUnix)
+            {
+                return;
+            }
+
+            IntPtr hdev = SetupApi.SetupDiGetClassDevs(ref SetupApi.GUID_DEVICE_BATTERY, IntPtr.Zero, IntPtr.Zero, SetupApi.DIGCF_PRESENT | SetupApi.DIGCF_DEVICEINTERFACE);
+            if (SetupApi.INVALID_HANDLE_VALUE != hdev)
+            {
+                // Limit battery search to 100 batteries
+                for (uint idev = 0; idev < 100; idev++)
+                {
+                    SetupApi.SP_DEVICE_INTERFACE_DATA did = default;
+                    did.cbSize = (uint)Marshal.SizeOf(typeof(SetupApi.SP_DEVICE_INTERFACE_DATA));
+
+                    if (SetupApi.SetupDiEnumDeviceInterfaces(hdev,
+                                                             IntPtr.Zero,
+                                                             ref SetupApi.GUID_DEVICE_BATTERY,
+                                                             idev,
+                                                             ref did))
+                    {
+                        
+                        SetupApi.SetupDiGetDeviceInterfaceDetail(hdev,
+                                                                 did,
+                                                                 IntPtr.Zero,
+                                                                 0,
+                                                                 out uint cbRequired,
+                                                                 IntPtr.Zero);
+                        if (Marshal.GetLastWin32Error() == SetupApi.ERROR_INSUFFICIENT_BUFFER)
+                        {
+                            IntPtr pdidd = Kernel32.LocalAlloc(Kernel32.LPTR, cbRequired);
+                            SetupApi.SP_DEVICE_INTERFACE_DETAIL_DATA didd = SetupApi.SP_DEVICE_INTERFACE_DETAIL_DATA.Default;
```

Comment:
You don't seem to be doing anything with SP_DEVICE_INTERFACE_DETAIL_DATA, might as well remove this.

Reply:
Also force SetupDiGetDeviceInterfaceDetail to unicode variant so you can remove CharSet.Auto stuff.

Reply:
> You don't seem to be doing anything with SP_DEVICE_INTERFACE_DETAIL_DATA, might as well remove this.

I've removed the "didd" variable, but it's necessary to do the conversion from SP_DEVICE_INTERFACE_DETAIL_DATA to IntPtr.

> Also force SetupDiGetDeviceInterfaceDetail to unicode variant so you can remove CharSet.Auto stuff.

Done

Reply:
Thanks @EMN-CSharp, I have reformatted it a bit, can you check if everything is still working? Worked OK here but one more is always good, then I'll merge.

Reply:
Everything worked OK here:
![LHMNow](https://user-images.githubusercontent.com/84488929/152710427-8a3b0fb6-8bb3-41d4-af18-fe02a0025ef7.PNG)